cool-core/src/main/java/com/nus/cool/core/cohort/refactor/CohortProcessor.java

```diff
@@ -0,0 +1,258 @@
+        }
+    }
+
+    /**
+     * In this section, we load the tuple which is a inner property.
+     * We left the process logic in procesTuple function.
+     * 
+     * @param chunk
+     * @param hashMapperBySchema
+     */
+    private void processDataChunk(ChunkRS chunk, HashMap<String, String[]> hashMapperBySchema) {
+        for (int i = 0; i < chunk.getRecords(); i++) {
+            // load data into tuple
+            for (String schema : this.projectedSchemaSet) {
+                // if the value is segment type, we should convert it to String from globalId
+                if(hashMapperBySchema.containsKey(schema)){
+                    int globalId = chunk.getField(schema).getValueByIndex(i);
+                    String v = hashMapperBySchema.get(schema)[globalId];
+                    tuple.loadAttr(v, schema);
+                } else {
+                    tuple.loadAttr(chunk.getField(schema).getValueByIndex(i), schema);
+                }
+            }
+            // now this tuple is loaded
+            this.processTuple();
+        }
+    }
+
+    /**
+     * process the inner tuple
+     */
+    private void processTuple() {
+        // For One Tuple, we firstly get the userId, and ActionTime
+        String userId = (String) tuple.getValueBySchema(this.UserIdSchema);
+        LocalDateTime actionTime = DateUtils.createCalender((int)tuple.getValueBySchema(this.ActionTimeSchema));
+        // check whether its birthEvent is selected
+        if (!this.birthSelector.isUserSelected(userId)) {
+            // if birthEvent is not selected
+            this.birthSelector.selectEvent(userId, actionTime, this.tuple);
+        } else {
+            // the birthEvent is selected
+
+            // do time_diff to generate age
+            // get the BirthEvent Date
+            LocalDateTime birthTime = this.birthSelector.getUserBirthEventDate(userId);
+            int age = this.ageSelector.generateAge(birthTime, actionTime);
+            if (age == AgeSelection.DefaultNullAge) {
+                // age is outofrange
+                return;
+            }
+
+            // extract the cohort this tuple belong to
+            String cohortName = this.cohortSelector.selectCohort(this.tuple);
+            if (cohortName == null) {
+                // cohort is outofrange
+                return;
+            }
+
+            if (!this.valueSelector.IsSelected(this.tuple)) {
+                // value outofrange
+                return;
+            }
+            // Pass all above filter, we can store value into CohortRet
+            // get the temporay result for this CohortGroup and this age
+            System.out.println("[Update Cohort Result]: cohortName:" + cohortName + "\tage:"+ age);
+            RetUnit ret = this.result.get(cohortName, age);
+            // update
+            this.valueSelector.updateRetUnit(ret, tuple);
+        }
+    }
+
+    private boolean checkMetaChunk(MetaChunkRS meta) {
```

Comment:
Has not implemented MetaChunk filtering?

Reply:
Next-phase task